### ISSUE TITLE
Issue with multiple hostnames in the same server

### ISSUE THREAD

**Issue by yozank** [hazelcast/hazelcast#15722]
2 Hazelcast instances with different group name configurations are put in the same box. 

When the member addresses are configured via IP address, they simply form separate clusters. 
However, if the member addresses are configured via host names, second instances shuts down instead of forming a separate cluster. 

Below is the explanation of the issue by Joseph: 

The problem seems to be located in the way how we store connections in the `TcpIpEndpointManager`. It’s a `Map` where the key is an `Address` instance

```
    final ConcurrentHashMap<Address, TcpIpConnection> connectionsMap =
            new ConcurrentHashMap<Address, TcpIpConnection>(100);
```

The bind request successfully finishes and the BindHandler registers the connection in the map. But it uses the address ‘aliases’ provided by the remote side (e.g. member1):

* https://github.com/hazelcast/hazelcast/blob/v3.12.1/hazelcast/src/main/java/com/hazelcast/nio/tcp/BindHandler.java#L158-L159
* https://github.com/hazelcast/hazelcast/blob/v3.12.1/hazelcast/src/main/java/com/hazelcast/nio/tcp/BindHandler.java#L194-L201

Calling the method `getOrConnect` ([source](https://github.com/hazelcast/hazelcast/blob/v3.12.1/hazelcast/src/main/java/com/hazelcast/nio/tcp/TcpIpEndpointManager.java#L180-L188)) then tries to find the existing connection on the map, but the hostname is from the current view (e.g. `member2`)

```
    public TcpIpConnection getOrConnect(final Address address, final boolean silent) {
        TcpIpConnection connection = connectionsMap.get(address);
        if (connection == null && networkingService.isLive()) {
            if (connectionsInProgress.add(address)) {
                connector.asyncConnect(address, silent);
            }
        }
        return connection;
    }
```

Unfortunately, the `Address`’s `equals` and `hashCode` methods work with hostnames and not with its IP representation:

```
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Address)) {
            return false;
        }
        final Address address = (Address) o;
        return port == address.port && this.type == address.type && this.host.equals(address.host);
    }
```

As a result the connection is not found by the `TcpIpEndpointManager` even the bind request was successful before and the connection is on the map. And the new connection is tried.

**Workaround**: use the same hostname in both configurations

**Possible fix**: use `InetSockAddress` instead host and port comparison in the Address.hashCode() and equals()

**Comment by sertugkaya** [comment]
Resolves ZD 5502. Customer was told this will be included in 3.12.5, currently they have a workaround.

**Comment by kwart** [comment]
We realized there is no simple solution for this member's misconfiguration. We will need to work on improving the documentation to cover it better. (hazelcast/hazelcast-reference-manual#789)

Socket addresses (host:port pairs) from different clusters should not overlap. And if they do, then member list should be filled by pairs belonging to the correct cluster in the TcpIp join configuration.

**Comment by kwart** [comment]
Moving the issue to version 4.1.

We will need to rework TCP connection mappings. Currently, the mapping is based on our `Address` object which proved to be error-prone. The plan is to use mapping based on member unique identifier (UUID). As this change is going to be an intrusive one we will not cover it in 4.0 in this stage. We only do a member protocol preparation work in 4.0 - i.e. include UUID within initial member-to-member communication (`BindMessage`).

**Comment by mmedenjak** [comment]
@kwart @jbartok is this issue somehow affected by the changes in https://github.com/hazelcast/hazelcast/pull/18591?

**Comment by vbekiaris** [comment]
moving to 5.1 as it might be addressed in the context of #19883 /cc @ufukyilmaz
